packages/mui-material/src/Badge/Badge.js

```diff
@@ -238,65 +246,54 @@ const Badge = React.forwardRef(function Badge(inProps, ref) {
     variant = variantProp,
   } = invisible ? prevProps : props;
 
-  const ownerState = { ...props, anchorOrigin, invisible, color, overlap, variant };
-  const classes = useUtilityClasses(ownerState);
+  const displayValue = variant !== 'dot' ? displayValueFromHook : undefined;
 
-  let displayValue;
+  const ownerState = {
+    ...props,
+    badgeContent,
+    invisible,
+    max,
+    displayValue,
+    showZero,
+    anchorOrigin,
+    color,
+    overlap,
+    variant,
+  };
 
-  if (variant !== 'dot') {
-    displayValue =
-      badgeContentProp && Number(badgeContentProp) > max ? `${max}+` : badgeContentProp;
-  }
+  const classes = useUtilityClasses(ownerState);
 
   // support both `slots` and `components` for backward compatibility
-  const RootSlot = slots?.root ?? components.Root ?? BadgeRoot;
+  const RootSlot = component ?? slots?.root ?? components.Root ?? BadgeRoot;
```

Comment:
The `component` prop, if present, should go to the `as` prop of the Root slot, not replace the slot completely.